crates/core/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+use quickjs_wasm_rs::Context;
+use std::alloc::{alloc, dealloc, Layout};
+use std::io;
+use std::ptr::copy_nonoverlapping;
+use std::slice;
+use std::str;
+
+mod globals;
+
+// Unlike C's realloc, zero-length allocations need not have
+// unique addresses, so a zero-length allocation may be passed
+// in and also requested, but it's ok to return anything that's
+// non-zero to indicate success.
+const ZERO_SIZE_ALLOCATION_PTR: *mut u8 = 1 as _;
+
+/// Compiles JS source code to QuickJS bytecode.
+///
+/// Returns a pointer to a byte array containing the bytecode.
+///
+/// # Arguments
+///
+/// * `js_src_ptr` - A pointer to the start of a byte array containing UTF-8 JS source code
+/// * `js_src_len` - The length of the byte array containing JS source code
+/// * `bytecode_len_ptr` - A pointer to where to write the length of the returned bytecode byte array
+///
+/// # Safety
+///
+/// * `js_src_ptr` must reference a valid array of unsigned bytes of `js_src_len` length
+/// * `bytecode_len` must be a pointer allocated by `canonical_abi_realloc` with an `alignment` of
+///   4 and a `new_size` of 1
+#[export_name = "compile-src"]
+pub unsafe extern "C" fn compile_src(
+    js_src_ptr: *const u8,
+    js_src_len: usize,
+    bytecode_len_ptr: *mut u32,
+) -> *const u8 {
+    let context = Context::default();
+    let js_src = str::from_utf8(slice::from_raw_parts(js_src_ptr, js_src_len)).unwrap();
+    let bytecode = context.compile_global("function.mjs", js_src).unwrap();
+    *bytecode_len_ptr = bytecode.len().try_into().unwrap();
+    // We need the bytecode buffer to live longer than this function so it can be read from memory
+    Box::leak(bytecode.into_boxed_slice()).as_ptr()
+}
+
+/// Evaluates QuickJS bytecode
+///
+/// # Safety
+///
+/// * `bytecode_ptr` must reference a valid array of unsigned bytes of `bytecode_len` length
+#[export_name = "eval-bytecode"]
```

Comment:
Underscores were used on the memory functions because those are defined as part of the component model. Honestly the only reason I used dashes for these was that was what a prototype branch used and it carried over to this branch from not revisiting that. I've put through an update to use underscores.